File: packages/shared/src/schemas.ts
import { z } from "zod";

/**
 * Runtime validators mirroring `types.ts`. Use these at every trust
 * boundary — API route handlers, webhook receivers, and MCP tool
 * responses — so malformed telemetry fails fast instead of silently
 * corrupting the dashboard.
 */

export const severitySchema = z.enum(["info", "warning", "critical"]);

export const alertStatusSchema = z.enum(["open", "acknowledged", "resolved"]);

export const spanKindSchema = z.enum([
  "llm",
  "tool",
  "retrieval",
  "agent",
  "internal",
]);

export const spanStatusSchema = z.enum(["ok", "error", "timeout"]);

export const spanSchema = z.object({
  spanId: z.string(),
  traceId: z.string(),
  parentSpanId: z.string().nullable(),
  name: z.string(),
  kind: spanKindSchema,
  status: spanStatusSchema,
  startTime: z.string(),
  durationMs: z.number().nonnegative(),
  attributes: z.record(z.string(), z.union([z.string(), z.number(), z.boolean()])),
  errorMessage: z.string().optional(),
});

export const traceSchema = z.object({
  traceId: z.string(),
  agentName: z.string(),
  environment: z.string(),
  startTime: z.string(),
  durationMs: z.number().nonnegative(),
  status: spanStatusSchema,
  spanCount: z.number().int().nonnegative(),
  tokenCount: z.number().int().nonnegative().optional(),
  costUsd: z.number().nonnegative().optional(),
  spans: z.array(spanSchema),
});

export const traceSummarySchema = z.object({
  traceId: z.string(),
  agentName: z.string(),
  environment: z.string(),
  startTime: z.string(),
  durationMs: z.number().nonnegative(),
  status: spanStatusSchema,
  spanCount: z.number().int().nonnegative(),
});

export const alertSchema = z.object({
  alertId: z.string(),
  title: z.string(),
  severity: severitySchema,
  status: alertStatusSchema,
  agentName: z.string(),
  ruleName: z.string(),
  traceId: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  summary: z.string(),
});

export const metricPointSchema = z.object({
  timestamp: z.string(),
  value: z.number(),
});

export const metricSeriesSchema = z.object({
  metric: z.string(),
  unit: z.string(),
  points: z.array(metricPointSchema),
});

export const diagnosisSchema = z.object({
  diagnosisId: z.string(),
  traceId: z.string(),
  alertId: z.string().nullable(),
  createdAt: z.string(),
  rootCause: z.string(),
  confidence: z.number().min(0).max(1),
  suggestedFix: z.string(),
  relatedSpanIds: z.array(z.string()),
});

export const signozWebhookPayloadSchema = z.object({
  ruleId: z.string(),
  ruleName: z.string(),
  severity: severitySchema,
  status: z.enum(["firing", "resolved"]),
  labels: z.record(z.string(), z.string()),
  annotations: z.record(z.string(), z.string()),
  startsAt: z.string(),
  endsAt: z.string().optional(),
});

export const dashboardOverviewSchema = z.object({
  activeAgents: z.number().int().nonnegative(),
  tracesLast24h: z.number().int().nonnegative(),
  openAlerts: z.number().int().nonnegative(),
  criticalAlerts: z.number().int().nonnegative(),
  avgLatencyMs: z.number().nonnegative(),
  errorRatePct: z.number().min(0).max(100),
});

File: backend/models/schemas.py
"""
Pydantic models mirroring packages/shared/src/types.ts.

Keep these two files in sync by hand for now. If drift becomes a problem,
a later phase can generate one from the other (e.g. datamodel-code-generator
from a shared JSON Schema).
"""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class Severity(str, Enum):
    info = "info"
    warning = "warning"
    critical = "critical"


class AlertStatus(str, Enum):
    open = "open"
    acknowledged = "acknowledged"
    resolved = "resolved"


class SpanKind(str, Enum):
    llm = "llm"
    tool = "tool"
    retrieval = "retrieval"
    agent = "agent"
    internal = "internal"


class SpanStatus(str, Enum):
    ok = "ok"
    error = "error"
    timeout = "timeout"


class Span(BaseModel):
    span_id: str = Field(alias="spanId")
    trace_id: str = Field(alias="traceId")
    parent_span_id: Optional[str] = Field(default=None, alias="parentSpanId")
    name: str
    kind: SpanKind
    status: SpanStatus
    start_time: str = Field(alias="startTime")
    duration_ms: float = Field(alias="durationMs")
    attributes: dict[str, str | float | bool] = Field(default_factory=dict)
    error_message: Optional[str] = Field(default=None, alias="errorMessage")

    class Config:
        populate_by_name = True


class Trace(BaseModel):
    trace_id: str = Field(alias="traceId")
    agent_name: str = Field(alias="agentName")
    environment: str
    start_time: str = Field(alias="startTime")
    duration_ms: float = Field(alias="durationMs")
    status: SpanStatus
    span_count: int = Field(alias="spanCount")
    token_count: Optional[int] = Field(default=None, alias="tokenCount")
    cost_usd: Optional[float] = Field(default=None, alias="costUsd")
    spans: list[Span] = Field(default_factory=list)

    class Config:
        populate_by_name = True


class Alert(BaseModel):
    alert_id: str = Field(alias="alertId")
    title: str
    severity: Severity
    status: AlertStatus
    agent_name: str = Field(alias="agentName")
    rule_name: str = Field(alias="ruleName")
    trace_id: Optional[str] = Field(default=None, alias="traceId")
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    summary: str

    class Config:
        populate_by_name = True


class SignozWebhookPayload(BaseModel):
    rule_id: str = Field(alias="ruleId")
    rule_name: str = Field(alias="ruleName")
    severity: Severity
    status: str  # "firing" | "resolved"
    labels: dict[str, str] = Field(default_factory=dict)
    annotations: dict[str, str] = Field(default_factory=dict)
    starts_at: str = Field(alias="startsAt")
    ends_at: Optional[str] = Field(default=None, alias="endsAt")

    class Config:
        populate_by_name = True


class Diagnosis(BaseModel):
    diagnosis_id: str = Field(alias="diagnosisId")
    trace_id: str = Field(alias="traceId")
    alert_id: Optional[str] = Field(default=None, alias="alertId")
    created_at: str = Field(alias="createdAt")
    root_cause: str = Field(alias="rootCause")
    confidence: float
    suggested_fix: str = Field(alias="suggestedFix")
    related_span_ids: list[str] = Field(default_factory=list, alias="relatedSpanIds")
    impact: str = Field(default="Unknown impact")
    next_steps: list[str] = Field(default_factory=list, alias="nextSteps")

    class Config:
        populate_by_name = True

